File: orientedDisign/library/notification_service.py
from typing import Dict


class NotificationService:
    def __init__(self):
        self._subscribers: Dict[str, object] = {}

    def register(self, user):
        self._subscribers[user.user_id] = user

    def unregister(self, user):
        self._subscribers.pop(user.user_id, None)

    def notify_user(self, user_id: str, message: str):
        user = self._subscribers.get(user_id)
        if user:
            user.notify(message)

    def notify_overdues(self, transactions, library_system):
        # transactions: iterable of BorrowTransaction
        for t in transactions:
            if t.overdue and not t.returned:
                user = self._subscribers.get(t.user_id)
                book = library_system.books.get(t.book_id)
                if user and book:
                    user.notify(f"Book '{book.title}' is overdue. Transaction: {t.transaction_id}")

File: orientedDisign/js_demo/notificationService.js
class NotificationService {
  constructor() {
    this.subscribers = new Map();
  }

  register(user) {
    this.subscribers.set(user.userId, user);
  }

  unregister(user) {
    this.subscribers.delete(user.userId);
  }

  notifyUser(userId, message) {
    const user = this.subscribers.get(userId);
    if (user) user.notify(message);
  }

  notifyOverdues(transactions, librarySystem) {
    for (const t of transactions) {
      if (t.overdue && !t.returned) {
        const user = this.subscribers.get(t.userId);
        const book = librarySystem.books.get(t.bookId);
        if (user && book) {
          user.notify(`Book '${book.title}' is overdue. Transaction: ${t.transactionId}`);
        }
      }
    }
  }
}

module.exports = NotificationService;
